### PR TITLE
[SPARK-41600][SPARK-41623][SPARK-41612][CONNECT] Implement Catalog.cacheTable, isCached and uncache

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR adds three API below to Spark Connect
- `Catalog.isCached`
- `Catalog.cacheTable`
- `Catalog uncacheTable`

### Why are the changes needed?

These were not added because of the design concern (in its behaviour). However, we should provide the same API compatibility and behaivours with the regular PySpark in any event. So these are proposed back.

### Does this PR introduce _any_ user-facing change?

No to end users.
Yes to the dev because it adds three new API in Spark Connect.

### How was this patch tested?

Unittests were added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @zhengruifeng, @grundprinzip @ueshin FYI

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/tests/test_dataframe.py [hunk 1]
```diff
@@ -1041,17 +1041,17 @@ def test_cache(self):
...
             )
             self.assertRaisesRegex(
-                AnalysisException,
+                Exception,
                 "does_not_exist",
                 lambda: spark.catalog.cacheTable("does_not_exist"),
             )
             self.assertRaisesRegex(
-                AnalysisException,
+                Exception,
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_dataframe.py:1054]: Should change back once https://github.com/apache/spark/pull/39882 is merged.

File: python/pyspark/sql/catalog.py [hunk 2]
```diff
@@ -956,7 +959,7 @@ def isCached(self, tableName: str) -> bool:
 
         Throw an analysis exception when the table does not exist.
 
-        >>> spark.catalog.isCached("not_existing_table")
+        >>> spark.catalog.isCached("not_existing_table")  # doctest: +SKIP
```
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py:962]: Should enable this back too. I will fix them all once https://github.com/apache/spark/pull/39882 is merged.

Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py:962]: Should enable this back too once https://github.com/apache/spark/pull/39882 is merged.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs)

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Comment by HyukjinKwon** [comment]
Thank you guys.

**Pull request closed without merging**